#Initialization
#https://nbisweden.github.io/workshop-scRNAseq/labs/compiled/seurat/seurat_07_spatial.html
#https://rdrr.io/github/dmcable/RCTD/f/vignettes/spatial-transcriptomics.Rmd
```{r}
library(Seurat)
library(Matrix)
library(ggplot2)
library(patchwork)
library(dplyr)
library(readxl)
library(scatterpie)


prefix = "Visium.MetsGBM.CombinedAnalysis.Aug2022"
sampleTable = "VisiumSamplelist.xlsx"

dirs <- read_xlsx(sampleTable)
IDs = as.character(dirs$ID)
Samplelist = IDs

```


#Read the 10x output files
```{r}

ob.list = list()
objectlist = c()
Varlist = c()
#drawPlot=0
#i=1

for (i in 1:length(IDs)){
    sampleID <- IDs[i]
  	sample <- as.character(dirs[i,2])
  	path1 <- paste(sample,sampleID,"outs",sep="/")
  	print(path1)
  	
  	data <- Load10X_Spatial(path1)
  	
  	#remove cell with umi sum 0, which cause error when doing scTransform
  	if(any(colSums(data@assays$Spatial@counts)==0)){ 
  	  test = as.list(colSums(data@assays$Spatial@counts))
  	  cells_include = names(test[test>0])
  	  data = subset(data,cells = cells_include )
  	}
  	
  	#data <- RenameCells(object = data[["Spatial"]], add.cell.id = sampleID)
  	
  	### Data preprocessing ###########################################################
  	#We first need to normalize the data in order to account for variance in sequencing depth        #across data points. 
    plot1 <- VlnPlot(data, features = "nCount_Spatial", pt.size = 0.1) + NoLegend()
    plot2 <- SpatialFeaturePlot(data,features = "nCount_Spatial")+theme(legend.position="right")
  	
    outfile = paste(prefix,sampleID,"Count.png",sep="_")
    png(filename = outfile,width = 800, height = 1200)
    t <- wrap_plots(plot1, plot2)
    print(t)
  	dev.off()
  	
    plot1 <- VlnPlot(data, features = "nFeature_Spatial", pt.size = 0.1) + NoLegend()
    plot2 <- SpatialFeaturePlot(data,features ="nFeature_Spatial")+theme(legend.position="right")
  	
    outfile = paste(prefix,sampleID,"Feature.png",sep="_")
    png(filename = outfile,width = 800, height = 1200)
    t <- wrap_plots(plot1, plot2)
    print(t)
  	dev.off()  	

    data.tran <- SCTransform(data, assay = "Spatial", method = "poisson") #sctransform (Hafemeister and Satija, Genome Biology 2019) builds regularized negative binomial models of gene expression in order to account for technical artifacts while preserving biological variance.

    data.tran$ID = sampleID
    ob.list = append(ob.list,data.tran)
    Varlist = c(Varlist, VariableFeatures(data.tran))

    objectlist = c(objectlist,sampleID)
    gc()
}  	


```


#integrate data run PCA and umap
```{r}

library(future)
library(future.apply)
plan("multiprocess", workers = 4)
options(future.globals.maxSize = 20 * 1024^3)


st.features = SelectIntegrationFeatures(ob.list, nfeatures = 3000, verbose = FALSE)
ob.list <- PrepSCTIntegration(object.list = ob.list, anchor.features = st.features, verbose = FALSE)

int.anchors <- FindIntegrationAnchors(object.list = ob.list, normalization.method = "SCT", verbose = FALSE, anchor.features = st.features)

data.merge <- IntegrateData(anchorset = int.anchors, normalization.method = "SCT", verbose = FALSE)

#rm(int.anchors, ob.list)
gc()

data.merge <- RunPCA(data.merge, verbose = FALSE)
data.merge <- FindNeighbors(data.merge, dims = 1:30)
data.merge <- RunUMAP(data.merge, dims = 1:30)
data.merge <- FindClusters(data.merge, resolution = 0.3, verbose = FALSE)

Idents(data.merge) <- "seurat_clusters"

outfile = paste(prefix,"_UMAP.Res0.3.png",sep="")
png(filename = outfile,width = 800, height = 800)
DimPlot(data.merge, reduction = "umap", group.by = c("ident"), label = TRUE, label.size = 8)
dev.off()

outfile = paste(prefix,"_SpatialUMAP.Res0.3.png",sep="")
png(filename = outfile,width = 1600, height = 400)
SpatialDimPlot(data.merge)
dev.off()


```


#plot and find cluster gene markers
```{r}

DefaultAssay(data.merge) <- "Spatial"
Idents(data.merge) <- "seurat_clusters"

cluster.markers<- FindAllMarkers(object = data.merge, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0)
outfile = paste(prefix,"_ClusterDiffGene.Res0.3.csv",sep="")
write.csv(cluster.markers, file=outfile)


#DefaultAssay(data.merge) <- "Spatial"
DefaultAssay(data.merge) <- "SCT"

genelist = c("PTPRC","CD3D","FOXP3","CD14","FCN1","CLEC9A","CLEC10A","TMEM119","HLA-DRA","MRC1","LYVE1","CD79A","PECAM1","ACTA2","RGS5","CLDN5","FN1","GFAP","SOX2","OLIG1","OLIG2","NEUROD1","EGFR","PMEL","MITF","EPCAM","MKI67")

genelist = c("PMEL","EPCAM","SOX2","MKI67","GFAP")

# outfile = paste(prefix,"SpatialGene.MainMarker.png",sep="_")
# png(filename = outfile,width = 5000, height = 5000)
outfile = paste(prefix,"SpatialGene.MainMarker.tiff",sep="_")
tiff(file=outfile, width=5000, height=5000, res=600)
SpatialFeaturePlot(data.merge, features = genelist, pt.size.factor=1.6)
dev.off()


saveRDS(data.merge,file=paste(prefix, ".rds", sep=""))
#data.merge = readRDS(file=paste(prefix, ".rds", sep=""))

```
